File: api/java-client/src/main/java/screencastnarrator/Highlight.java
package screencastnarrator;

import com.microsoft.playwright.ElementHandle;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

import java.util.Arrays;
import java.util.List;

public class Highlight {

    private static final int ACK_POLLING_INTERVAL_MS = 40;

    private static final String REMOVE_WRAPPER_JS = "document.getElementById('_e2e_multi_highlight')?.remove();";

    private Highlight() {}

    public static void highlight(Page page, Locator[] locators, SharedConfig config, CdpVideoRecorder recorder) {
        locators[0].evaluate(config.highlight().getScrollJs());
        page.evaluate(config.highlight().getScrollWaitJs());

        List<ElementHandle> handles = Arrays.stream(locators)
                .map(Locator::elementHandle)
                .toList();

        page.evaluate(config.highlight().getCombineJs(), handles);
        page.locator("#_e2e_multi_highlight").evaluate(config.resolvedDrawJs());
        waitWithAcks(page, config.highlight().getAnimationSpeedMs() + config.highlight().getDrawWaitMs(), recorder);

        page.evaluate(config.highlight().getRemoveJs());
        page.evaluate(REMOVE_WRAPPER_JS);
        waitWithAcks(page, config.highlight().getRemoveWaitMs(), recorder);
    }

    public static void highlight(Page page, Locator locator, SharedConfig config, CdpVideoRecorder recorder) {
        highlight(page, new Locator[]{locator}, config, recorder);
    }

    public static void highlight(Page page, Locator locator, SharedConfig config) {
        highlight(page, locator, config, null);
    }

    static void waitWithAcks(Page page, int totalMs, CdpVideoRecorder recorder) {
        if (recorder == null || totalMs <= 0) {
            page.waitForTimeout(totalMs);
            return;
        }
        int elapsed = 0;
        int interval = ACK_POLLING_INTERVAL_MS;
        while (elapsed < totalMs) {
            int wait = Math.min(interval, totalMs - elapsed);
            page.waitForTimeout(wait);
            elapsed += wait;
            recorder.ackLatestFrame();
        }
    }
}

File: api/python-client/src/screencast_narrator_client/highlight.py
"""Element highlighting: draws animated highlights around Playwright locators."""

from __future__ import annotations

from screencast_narrator_client.shared_config import SharedConfig

_REMOVE_WRAPPER_JS = "document.getElementById('_e2e_multi_highlight')?.remove();"


def highlight(page, locators: list, config: SharedConfig) -> None:
    hl = config.highlight
    locators[0].evaluate(config.resolved_scroll_js)
    page.evaluate(config.resolved_scroll_wait_js)

    handles = [loc.element_handle() for loc in locators]
    page.evaluate(config.resolved_combine_js, handles)
    page.locator("#_e2e_multi_highlight").evaluate(config.resolved_draw_js)
    page.wait_for_timeout(hl.animation_speed_ms + hl.draw_wait_ms)

    page.evaluate(config.resolved_remove_js)
    page.evaluate(_REMOVE_WRAPPER_JS)
    page.wait_for_timeout(hl.remove_wait_ms)
